metamodel-export/src/main/kotlin/org/modelix/metamodel/export/MPSModelExporter.kt

```diff
@@ -0,0 +1,123 @@
+                    it,
+                    CONCEPTS.BaseConcept,
+                    false,
+                    arrayOf(),
+                )
+            }
+            .flatMap { SNodeOperations.getReferences(it) }
+            .mapNotNull { it.targetNode }.mapNotNull { it: SNode -> it.model }.distinct()
+            .toList()
+        for (referencedModel in referencedModels) {
+            collectDependencies(referencedModel, result)
+        }
+    }
+
+    private fun writeFiles(name: String?, nodeData: NodeData) {
+        val data = ModelData(null, nodeData)
+        val jsonFile = File(outputFolder, "$name.json")
+        jsonFile.writeText(data.toJson(), StandardCharsets.UTF_8)
+    }
+
+    fun exportModules(modules: List<SModule>): NodeData {
+        return exportModules(modules, null)
+    }
+
+    private fun exportModules(modules: List<SModule>, modelsToInclude: Set<SModel?>?): NodeData {
+        val modulesData: List<NodeData> = modules.map { it: SModule -> exportModule(it, modelsToInclude) }
+        val root = NodeData("", null, null, modulesData, emptyMap(), emptyMap())
+        return root
+    }
+
+    private fun exportModule(mpsModule: SModule, modelsToInclude: Set<SModel?>?): NodeData {
+        val models: Iterable<SModel> = mpsModule.models
+        val modelsData: List<NodeData> = models.filter { it: SModel? ->
+            modelsToInclude == null || modelsToInclude.contains(it)
+        }.map { it: SModel -> exportModel(it) }
+        val properties: MutableMap<String, String> = LinkedHashMap()
+        properties["id"] = mpsModule.moduleId.toString()
+        properties["name"] = mpsModule.moduleName ?: ""
+        return NodeData(
+            mpsModule.moduleReference.toString(),
+            "mps-module",
+            "modules",
+            modelsData,
+            properties,
+            emptyMap(),
+        )
+    }
+
+    private fun exportModel(mpsModel: SModel): NodeData {
+        val rootNodes: Iterable<SNode> = mpsModel.rootNodes
+        val rootNodeData: List<NodeData> = rootNodes.map { exportNode(it) }
+        val properties: MutableMap<String, String> = LinkedHashMap()
+        properties["id"] = mpsModel.modelId.toString()
+        properties["name"] = mpsModel.name.value
+        val data = NodeData(mpsModel.reference.toString(), "mps-model", "models", rootNodeData, properties, emptyMap())
+        return data
+    }
+
+    private fun exportNode(node: SNode): NodeData {
+        val id: String = node.reference.toString()
+        val concept: String = "mps:" + MetaIdHelper.getConcept(SNodeOperations.getConcept(node)).serialize()
+        val role = node.containmentLink?.name ?: "rootNodes"
+        val children: List<NodeData> = SNodeOperations.getChildren(node).map { exportNode(it) }
+        val properties: MutableMap<String, String> = LinkedHashMap()
+        val references: MutableMap<String, String> = LinkedHashMap()
+        properties["#mpsNodeId#"] = node.nodeId.toString()
```

Comment:
Maybe we should use `#originalRef#` here. But I don't know if it's worth it to change it here, since we wanted to replace this functionality with a part of the bulk-sync.

Reply:
This is the existing behavior. Changing that is out of scope for this PR. We will eventually remove this legacy code.

Reply:
https://issues.modelix.org/issue/MODELIX-842/Remove-model-export-feature-from-api-gen-gradle